File: gem5/src/mem/cxl/cxl_mem_dev.cc
#include "cxl_mem_dev.hh"
#include "mem/port.hh"

namespace gem5{
namespace memory{

Port &
CxlMemOnlyDev::getPort(const std::string &if_name, PortID idx)
{
    if (if_name == "to_internal_port") {
        return *to_internal_xbar_port;
    }
    return ClockedObject::getPort(if_name, idx);
}

CxlMemOnlyDev::CxlMemOnlyDev(const CxlMemOnlyDevParams &p)
    :ClockedObject(p), internal_xbar(p.InternalMemBus), external_xbar(p.ExternalBus),
    to_internal_xbar_port(NULL)
{
    to_internal_xbar_port = new MemSidePort("to_internal_port", InvalidPortID);
}

CxlMemOnlyDev::~CxlMemOnlyDev()
{
    delete to_internal_xbar_port;
}

}
}

File: gem5/src/mem/cxl/CxlMemDev.py
from m5.params import *
from m5.objects.AbstractMemory import AbstractMemory
from m5.objects.XBar import *
from m5.objects.ClockedObject import ClockedObject

class CxlMemInternalBus1(CoherentXBar):
    width = 512

class CxlMemOnlyDev(ClockedObject):
    type = 'CxlMemOnlyDev'
    cxx_header = "mem/cxl/cxl_mem_dev.hh"
    cxx_class = 'gem5::memory::CxlMemOnlyDev'

    abstract = False

    # 추후 더 나은 방식으로 수정 예정
    
    # 내부 용 메모리버스
    InternalMemBus = BaseXBar()
    # 외부 연결용 버스
    ExternalBus = BaseXBar()
    
    # 내부 메모리 버스에 연결된 메모리 장치 리스트
    #MemCtrls = []
    # 내부 컨트롤러 연결용
    to_internal_port = RequestPort("")
    from_internal_port = ResponsePort("")

    def connect_memctrls(self, memctrls):
        for i in memctrls:
            self.InternalMemBus.mem_side_ports = i
        self.InternalMemBus.cpu_side_ports = self.to_internal_port